osu.Game/Graphics/Containers/SectionsContainer.cs

```diff
@@ -177,36 +183,90 @@ protected override void UpdateAfterChildren()
+                return;
+
+            lastKnownScroll = scrollContainer.Current;
 
-            if (currentScroll != lastKnownScroll)
+            if (ExpandableHeader != null && FixedHeader != null)
             {
-                lastKnownScroll = currentScroll;
+                float offset = Math.Min(ExpandableHeader.LayoutSize.Y, scrollContainer.Current);
 
-                if (ExpandableHeader != null && FixedHeader != null)
-                {
-                    float offset = Math.Min(ExpandableHeader.LayoutSize.Y, currentScroll);
+                ExpandableHeader.Y = -offset;
+                FixedHeader.Y = -offset + ExpandableHeader.LayoutSize.Y;
+            }
 
-                    ExpandableHeader.Y = -offset;
-                    FixedHeader.Y = -offset + ExpandableHeader.LayoutSize.Y;
-                }
+            headerBackgroundContainer.Height = (ExpandableHeader?.LayoutSize.Y ?? 0) + (FixedHeader?.LayoutSize.Y ?? 0);
+            headerBackgroundContainer.Y = ExpandableHeader?.Y ?? 0;
 
-                headerBackgroundContainer.Height = (ExpandableHeader?.LayoutSize.Y ?? 0) + (FixedHeader?.LayoutSize.Y ?? 0);
-                headerBackgroundContainer.Y = ExpandableHeader?.Y ?? 0;
+            selectMostVisible();
+        }
 
-                float scrollOffset = FixedHeader?.LayoutSize.Y ?? 0;
-                Func<T, float> diff = section => scrollContainer.GetChildPosInContent(section) - currentScroll - scrollOffset;
+        /// <summary>
+        /// Changes <see cref="SelectedSection"/> to currently most visible section
+        /// </summary>
+        private void selectMostVisible()
+        {
+            // reset last clicked section because user started scrolling themselves
+            if (scrollContainer.UserScrolling)
+                lastClickedSection = null;
 
-                if (scrollContainer.IsScrolledToEnd())
-                {
-                    SelectedSection.Value = Children.LastOrDefault();
-                }
-                else
-                {
-                    SelectedSection.Value = Children.TakeWhile(section => diff(section) <= 0).LastOrDefault()
-                                            ?? Children.FirstOrDefault();
-                }
+            // we are scrolled all the way to the bottom
+            // select the section user clicked for or the very last section
+            if (Precision.AlmostBigger(scrollContainer.Current, scrollContainer.ScrollableExtent))
+            {
+                SelectedSection.Value = lastClickedSection as T ?? Children.LastOrDefault();
+                return;
             }
+
+            var sortedByVisibility = new List<T>(Children);
+            sortedByVisibility.Sort(compareVisibility);
+            SelectedSection.Value = sortedByVisibility.FirstOrDefault();
+        }
+
+        private int compareVisibility(T a, T b)
+        {
+            // note: the decision to prioritize higher sections and differentiate scroll direction
```

Comment:
honestly this feels wrong, and doesn't match stable. why aren't we centering the target instead? will that simplify things?

Reply:
It won't really be more simple, it will just be different. There will be other edge cases you would need to consider if the sections are scrolled to middle

Reply:
Actually, if we were to scroll to middle, then an algorithm other than visibility is probably the best. You would probably want to select the section intersecting with the middle line, which might be easier to do changing the old code. The edge case would come from big sections that won't fit on screen and would still need to be scrolled to the top, and from the bottom of the screen where there might be several sections that can never reach the middle but should still be highlighted if clicked on.

Reply:
Yeah, it would probably need to scroll the start of the section to have the start of the section sitting at around 25% of the vertical height of the settings container (rather than the precise centre of the section). I'm not sure if stable already does this or not.